### PR TITLE
linear_cross_entropy: support reduction='none' on the chunked path

### PR DESCRIPTION
**Pull request by pearu**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* __->__ #186113

----

- Adds reduction='none' via a separate op (_linear_cross_entropy_batch_chunked_no_reduction) whose forward returns the per-sample loss (N,). Unlike the scalar-reduction op it cannot precompute gradients in forward -- a vector loss makes grad_output per-sample, and storing the per-sample (N, V) softmax to defer the weight/bias grads would defeat chunking -- so backward recomputes the chunked grads, folding grad_output into the per-row weight (each grad term is linear in it, giving the correct grad_output-weighted VJP). Peak memory stays bounded to (batch_chunk_size, num_classes), and retain_graph works for free (recompute from save_for_backward inputs, no consumed buffers). The shared chunk loop is factored into _linear_cross_entropy_batch_chunked_accumulator, called by the scalar forward, the none forward, and the none backward; the scalar path's chunked loop is unchanged.

Implementation assisted by Claude.

Test Plan: python test/test_nn.py -k linear_cross_entropy

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### CODE REVIEW

**Comment by slayton58** [comment]
@claude review this PR, be exhaustive and find all issues, concentrating on subtle errors that a human might miss. Ignore small things, give me a concise summary of only the most significant problems.

**Comment by pearu** [comment]
Here's a response to the review:
1. Redundant output work — confirmed it's harmless (the in-place log_() at line 814 runs after the grad path reads softmax_denom at 808). The cost is two O(B_chunk) dot products per chunk vs the O(B_chunk·V·F) matmuls, so I'm leaving it: gating it would need a compute_output flag threaded through an already flag-heavy shared loop, for ~0 speedup.
2. No none+bias OpInfo samples — good catch, fixed. Added reduction='none' + linear_bias samples (bias, bias+weight, bias+ignore_index) to the chunked_none variant so the infra tests exercise the none bias path; numerics stay in the dedicated test_nn.py tests.
3. vmap fast path — agreed it's now unblocked for both ops. Leaving it as a follow-up rather than expanding this PR's scope; the per-sample loop is the safe baseline. I've updated the vmap rule's comment to note the blocker is resolved (the reduction='none' chunked op now exists) and to flag the opportunity so it's discoverable.

**Comment by pearu** [comment]
### Benchmark: chunked `reduction='none'` vs Liger (A100, bfloat16)

auto_dispatch_bfloat16_NVIDIA_A100_SXM4_80GB_none


Auto-dispatch (`acc_policy="auto"`, `acc_dtype=fp32`) chunked `reduction='none'` vs Liger's fused linear-cross-entropy, swept over `num_tokens` / `in_features` / `num_classes`, gradients referenced against fp64:

- **Memory — ~2× lower than Liger, widening with vocabulary:** 0.63 vs 1.33 GiB at `num_classes=32000`, 1.14 vs 2.61 GiB at 65536; lower on every point of every axis.
- **Time — Liger-class:** on par at the largest vocab (79 vs 80 ms at V=65536), ~13% slower at V=32000, slower on the token axis — the `none` backward recomputes the chunked grads (one extra logits matmul per chunk), since the per-sample loss can't be precomputed in forward.
- **Accuracy — tied, slightly better at large vocab:** grad rel-err is bf16-eps-level for both; at V=65536 this PR is 1.7e-3 vs Liger 3.1e-3 (the fp32 accumulation helps where the class reduction hurts most).

**Bottom line:** vs Liger, the chunked `reduction='none'` path delivers ~2× lower peak memory that scales with vocabulary, at roughly Liger-class speed (slightly slower on tokens, on par at the largest vocab) and equal-or-better bf16 accuracy.

<sub>One rough edge: at small `in_features` with huge `num_classes` the `aspect_ratio` heuristic over-chunks and the per-chunk recompute dominates (~67 ms at `in_features=1024`, `num_classes=32000`) — a follow-up auto-dispatch tuning item for the `none` path, not a correctness issue.</sub>

**Review by slayton58 — APPROVED** [review]
